authorizer/source.go

```diff
@@ -0,0 +1,149 @@
+	p, err := newSourcePermission(influxdb.ReadAction, orgID, id)
+	if err != nil {
+		return err
+	}
+
+	if err := IsAllowed(ctx, *p); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func authorizeWriteSource(ctx context.Context, orgID, id influxdb.ID) error {
+	p, err := newSourcePermission(influxdb.WriteAction, orgID, id)
+	if err != nil {
+		return err
+	}
+
+	if err := IsAllowed(ctx, *p); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DefaultSource checks to see if the authorizer on context has read access to the default source.
+func (s *SourceService) DefaultSource(ctx context.Context) (*influxdb.Source, error) {
+	src, err := s.s.DefaultSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := authorizeReadSource(ctx, src.OrganizationID, src.ID); err != nil {
+		return nil, err
+	}
+
+	return src, nil
+}
+
+// FindSourceByID checks to see if the authorizer on context has read access to the id provided.
+func (s *SourceService) FindSourceByID(ctx context.Context, id influxdb.ID) (*influxdb.Source, error) {
+	src, err := s.s.FindSourceByID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := authorizeReadSource(ctx, src.OrganizationID, id); err != nil {
+		return nil, err
+	}
+
+	return src, nil
+}
+
+// FindSources retrieves all sources that match the provided options and then filters the list down to only the resources that are authorized.
+func (s *SourceService) FindSources(ctx context.Context, opts influxdb.FindOptions) ([]*influxdb.Source, int, error) {
+	// TODO: we'll likely want to push this operation into the database since fetching the whole list of data will likely be expensive.
+	ss, _, err := s.s.FindSources(ctx, opts)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// This filters without allocating
+	// https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating
+	sources := ss[:0]
+	for _, src := range ss {
+		err := authorizeReadSource(ctx, src.OrganizationID, src.ID)
+		if err != nil && influxdb.ErrorCode(err) != influxdb.EUnauthorized {
+			return nil, 0, err
+		}
+
+		if influxdb.ErrorCode(err) == influxdb.EUnauthorized {
+			continue
+		}
+
+		sources = append(sources, src)
+	}
+
+	return sources, len(sources), nil
+}
+
+// CreateSource checks to see if the authorizer on context has write access to the global source resource.
+func (s *SourceService) CreateSource(ctx context.Context, src *influxdb.Source) error {
+	p, err := influxdb.NewPermission(influxdb.WriteAction, influxdb.SourcesResourceType, src.OrganizationID)
+	if err != nil {
+		return err
+	}
+
+	if err := IsAllowed(ctx, *p); err != nil {
+		return err
+	}
+
+	return s.s.CreateSource(ctx, src)
+}
+
+// UpdateSource checks to see if the authorizer on context has write access to the source provided.
+func (s *SourceService) UpdateSource(ctx context.Context, id influxdb.ID, upd influxdb.SourceUpdate) (*influxdb.Source, error) {
+	src, err := s.FindSourceByID(ctx, id)
```

Comment:
I think we want `s.s.FindSourceByID` here since otherwise you will require the read action in order to perform a write action.

Reply:
You're right.

BTW, i agree with you. I am fighting with myself since this morning. Trying to not refactor all this boilerplate into some common method or generating it.